# CS 4661: Introduction to Data Science
## Jay Tandel


### Problem: Cancer Diagnosis Using Machine Learning
In this homework, we work with a real dataset from UCI Database.


## A
Read the dataset file “Cancer.csv” (from github using the following command), and assign it to a Pandas DataFrame:

df  = pd.read_csv("https://github.com/mpourhoma/CS4661/raw/master/Cancer.csv") Check out the dataset. As you see, the dataset includes 9 numerical features. The last column is the binary label (“1” means it is a malignant cancer, “0” means it is a benign tumor). You will use all 9 features in this homework.

In [1]:
# Importing the required packages and libraries
import pandas as pd
import numpy as np
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.model_selection import cross_val_score
from sklearn.utils import resample 
from statistics import mode
from sklearn.ensemble import RandomForestClassifier

In [2]:
# Reading the csv file from web, and store it in panda DataFrame
df = pd.read_csv("https://github.com/mpourhoma/CS4661/raw/master/Cancer.csv")

In [3]:
#print the imported dataset
print(df)

     Clump_Thickness  Uniformity_of_Cell_Size  Uniformity_of_Cell_Shape  \
0                  5                        1                         1   
1                  5                        4                         4   
2                  3                        1                         1   
3                  6                        8                         8   
4                  4                        1                         1   
..               ...                      ...                       ...   
145                3                        1                         1   
146                9                        7                         7   
147               10                        8                         8   
148                1                        1                         1   
149                5                        1                         1   

     Marginal_Adhesion  Single_Epithelial_Cell_Size  Bare_Nuclei  \
0                    1         

In [4]:
# Creating the Feature Matrix for the dataset:

# create a python list of feature names that would like to pick from the dataset:
feature_cols = ['Clump_Thickness', 'Uniformity_of_Cell_Size', 
                'Uniformity_of_Cell_Shape', 'Marginal_Adhesion', 'Single_Epithelial_Cell_Size', 'Bare_Nuclei',
               'Bland_Chromatin', 'Normal_Nucleoli', 'Mitoses']

# use the above list to select the features from the original DataFrame
X = df[feature_cols] 


print(X)

     Clump_Thickness  Uniformity_of_Cell_Size  Uniformity_of_Cell_Shape  \
0                  5                        1                         1   
1                  5                        4                         4   
2                  3                        1                         1   
3                  6                        8                         8   
4                  4                        1                         1   
..               ...                      ...                       ...   
145                3                        1                         1   
146                9                        7                         7   
147               10                        8                         8   
148                1                        1                         1   
149                5                        1                         1   

     Marginal_Adhesion  Single_Epithelial_Cell_Size  Bare_Nuclei  \
0                    1         

In [5]:
# select label (the last column) from the DataFrame
y = df['Malignant_Cancer']

print(y)

0      0
1      0
2      0
3      0
4      0
      ..
145    0
146    1
147    1
148    0
149    0
Name: Malignant_Cancer, Length: 150, dtype: int64


## B
Use sklearn functions to split the dataset into testing and training sets with the following parameters: test_size=0.35, random_state=3.

In [6]:
# Randomly splitting the original dataset into training set and testing set

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.35, random_state=3)

In [7]:
# print the training set:
print(X_train)
print('\n')
print(y_train)

     Clump_Thickness  Uniformity_of_Cell_Size  Uniformity_of_Cell_Shape  \
50                 5                        1                         3   
73                 1                        1                         1   
34                10                        5                         5   
136               10                       10                        10   
19                10                       10                        10   
..               ...                      ...                       ...   
107                8                        8                         7   
21                 2                        5                         3   
0                  5                        1                         1   
131                2                        1                         1   
106                3                        4                         5   

     Marginal_Adhesion  Single_Epithelial_Cell_Size  Bare_Nuclei  \
50                   1         

In [8]:
# print the testing set:
print(X_test)
print('\n')
print(y_test)

     Clump_Thickness  Uniformity_of_Cell_Size  Uniformity_of_Cell_Shape  \
47                 5                        3                         4   
3                  6                        8                         8   
31                 5                        3                         3   
25                10                       10                        10   
15                10                        5                         5   
118                1                        1                         1   
89                 6                        5                         4   
6                  1                        1                         1   
103                8                        3                         5   
65                 5                       10                         6   
88                 1                        1                         1   
38                 5                        2                         3   
92                10     

## C
Use “Decision Tree Classifier” to predict Cancer based on the training/testing datasets that you built in part (b). Then, calculate and report the accuracy of your classifier.

Use this command to define your tree: my_DecisionTree = DecisionTreeClassifier(random_state=3).

In [9]:
# Instantiating object of DecisionTreeClassifier "class" with random_state=3:
my_DecisionTree = DecisionTreeClassifier(random_state=3)

# Training the model with train dataset we created using split
my_DecisionTree.fit(X_train, y_train)

# Testing the model with test dataset we created using split

# store the predicted values
my_DecisionTree_y_predict = my_DecisionTree.predict(X_test)

# Calculating the Accuracy of the prediction 

# we use accuracy_score function
my_DecisionTree_accuracy = accuracy_score(y_test, my_DecisionTree_y_predict)

print('Accuracy using Decision Tree is: ',my_DecisionTree_accuracy)

Accuracy using Decision Tree is:  0.8301886792452831


## D
 Now, we want to perform a new Ensemble Learning method called “Bagging” based on Voting on 19 decision tree classifiers.

Note: you should write your own code to perform Bagging (don’t use scikit-learn functions for Bagging!) To do so, you need to perform bootstrapping first. You can write a “for” loop with loop variable  i = 0…18.  

In each iteration of the loop, you have to:

1. make a bootstarp sample of the original “Training” Dataset (build in part(b)) with the size of bootstarp_size = 0.8*(Size of the original dataset). You can use the following command to generate a random bootstrap dataset (“i" is the variable of the loop, so the random_state changes in each iteration):

resample(X_train, n_samples = bootstarp_size , random_state=i , replace = True)

2. Define and train a new base decision tree classifier on this dataset in each iteration:

Base_DecisionTree = DecisionTreeClassifier(random_state=3).

3. Perform prediction using “this base classifier” on the original “Testing” Dataset X_test (build in part(b)), and save the prediction results for all testing samples.

After finishing the “for” loop, you should have 19 different predictions for EACH sample in your testing set. Then, Perform Voting to make the final decision on each data sample based on the votes of all 19 classifiers.

Finally, calculate and report the final accuracy of your Bagging (Voting) method.

Note: You do NOT need to calculate the accuracy of each one of the base classifiers in each round of the loop! You have to just perform Voting to make the final decision on each data sample, and then calculate the accuracy on the final results.

In [24]:
#print(X_train.size)
boot_y_predictions = []
for i in range(0,19):
    bootstarp_size = int (0.8*(X_train.shape[0]))
    boot_X_train, boot_y_train = resample(X_train, y_train, n_samples=bootstarp_size, random_state=i, replace = True)
    Base_DecisionTree = DecisionTreeClassifier(random_state=3)
    Base_DecisionTree.fit(boot_X_train, boot_y_train)
    Base_DecisionTree_y_predict = Base_DecisionTree.predict(X_test)
    boot_y_predictions.append(Base_DecisionTree_y_predict)
    print("Iteration: ",i ,"\n", Base_DecisionTree_y_predict)

Iteration:  0 
 [1 1 1 1 1 0 1 1 1 1 0 1 1 1 0 1 0 0 1 1 0 1 0 0 1 1 1 1 0 1 1 1 0 1 0 0 0
 0 1 0 0 0 1 1 0 0 1 1 1 1 0 1 1]
Iteration:  1 
 [1 1 1 1 1 0 1 1 1 1 0 1 1 1 0 1 0 0 1 1 0 1 0 0 1 1 1 1 0 1 1 1 0 0 0 0 0
 0 1 0 0 0 1 1 0 1 1 0 1 1 0 1 1]
Iteration:  2 
 [1 1 0 1 1 0 1 0 1 1 0 1 0 1 0 1 0 0 1 1 0 1 0 0 1 1 1 1 0 1 0 1 0 0 0 0 0
 0 1 0 0 0 1 0 0 0 1 1 1 1 0 1 1]
Iteration:  3 
 [0 1 0 1 1 0 1 0 1 1 0 0 0 1 0 1 0 0 1 1 0 1 0 0 1 1 1 1 0 1 1 1 0 0 0 0 0
 0 1 0 0 0 0 1 0 1 1 0 0 1 0 1 1]
Iteration:  4 
 [1 1 1 1 1 0 1 0 1 1 0 1 1 1 0 1 0 1 1 1 0 1 0 0 1 1 1 1 0 1 0 1 0 0 0 0 0
 0 1 0 0 0 1 1 0 0 1 1 1 1 0 1 1]
Iteration:  5 
 [1 1 1 1 1 0 1 0 1 1 0 0 0 1 0 1 0 0 1 1 0 1 0 0 1 1 1 1 0 1 1 1 0 0 0 0 0
 0 1 0 0 0 1 1 0 1 1 0 1 1 0 1 1]
Iteration:  6 
 [1 1 1 1 1 0 1 0 1 1 0 1 1 1 0 1 0 1 1 1 0 1 0 0 1 1 1 1 0 1 0 1 0 0 0 0 0
 0 1 0 0 0 1 1 0 0 1 1 1 1 0 1 1]
Iteration:  7 
 [0 1 1 1 1 0 1 0 1 1 0 0 0 1 0 1 0 0 1 1 0 1 0 0 1 1 1 1 0 1 1 1 0 0 0 0 0
 0 1 0 0 0 1 1 0 1 1 0 1 1 0 1 1]


In [21]:
# transform the predictions matrix
p_list = np.array(boot_y_predictions)
predictions_t = p_list.T

print(predictions_t)

final_prediction = []
for l in predictions_t:
    temp = mode(l)
    final_prediction.append(temp)
    #print(temp)

# Calculating the Accuracy of the prediction 

# we use accuracy_score function
Boot_DecisionTree_accuracy = accuracy_score(y_test, final_prediction)

print("Accuracy is: ", Boot_DecisionTree_accuracy)


[[1 1 1 ... 1 1 1]
 [1 1 1 ... 1 1 1]
 [1 1 0 ... 0 1 1]
 ...
 [0 0 0 ... 1 0 0]
 [1 1 1 ... 1 1 1]
 [1 1 1 ... 1 1 1]]
Accuracy is:  0.9056603773584906


## E
Use scikit-learn “Random Forest” classifier to predict Cancer based on the training/testing datasets that you built in part (b). Then, calculate and report the accuracy of your classifier. Use this command to import and define your classifier:

from   sklearn.ensemble    import    RandomForestClassifier

my_RandomForest = RandomForestClassifier(n_estimators = 19, bootstrap = True, random_state=3)

Similar to previous syntax, use my_RandomForest.fit  for training your random forest classifier and my_RandomForest.predict  for prediction.

In [15]:
# Instantiating object of RandomForestClassifier "class" with random_state=3:
my_RandomForest = RandomForestClassifier(n_estimators = 19, bootstrap = True, random_state=3)

# Training the model with train datase
my_RandomForest.fit(X_train, y_train)

# Testing the model with test dataset

# store the predicted values
my_RandomForest_y_predict = my_RandomForest.predict(X_test)

# Calculating the Accuracy of the prediction 

# we use accuracy_score function
my_RandomForest_accuracy = accuracy_score(y_test, my_RandomForest_y_predict)

print('Accuracy using Random Forest is: ', my_RandomForest_accuracy)

Accuracy using Random Forest is:  0.9245283018867925


## Answer Summary

### C
#### Accuracy using Decision Tree is:  0.8301886792452831

### D
#### Accuracy is:  0.9056603773584906

### E
#### Accuracy using Random Forest is:  0.9245283018867925